### PR TITLE
'diffexpr' can be an in-process function, fine-grained control over diff chunks

### LINKED ISSUE #17210: Add `idifffn` option to customize internal diff engine instead of external engine like `diffexpr`.
### Feature already in Vim?

_No response_

### Feature description

According to `:h diff-diffexpr`, the expression set to the `&diffexpr` option needs to read/write files on disk and perform external calls to some custom diffing program.
This works fine, but it defeats the very purpose of having integrated an [internal diff engine](https://vimways.org/2018/the-power-of-diff/) into vim.

On the other hand, the Lua API allows (great) explicit calls to the internal engine according to `:h vim.diff`, which would make something like the `&idifffn` option I'm suggesting possible.

Ideally, I would be able to write something like:
```vim
lua <<EOF
function MyDiff(opt)
    -- custom call(s) to vim.diff(a, b, opt)
   return result_string
end
EOF
set idifffn=MyDiff
```
.. or close, and have my diffing procedure customized without an external call and I/O accesses. What do you think? :)

### PR DESCRIPTION
**Pull request by iago-lito**
Hello, here is my first (PR here and my first) attempt to feature #17210 :) The `&diffexpr` option may optionally be prefixed with `->`. In this situation, no files are read or written on disk, and no external call is performed. Instead, the user is in charge of retrieving diffed buffers data as strings, then calculate the diff chunks.

I'm almost happy with the design, except that the "minimal" example supposed to use the new feature to reproduce vim's default behaviour contains a lot of boilerplate: transforming arrays of lines to strings, parsing `&diffopt` to forward every option to `vim.diff` individually. Would it not be easier if `vim.diff` defaulted to options specified in `&diffopt` instead? But maybe this belongs to another issue?

fix #17210

### CODE REVIEW

**Comment by lewis6991** [comment]
The title is a bit wrong. This doesn't impact `vim.diff`, it simply allows `diffexpr` to work completely in process and allows it to use `vim.diff` instead of being expected to run an external process.

File: src/nvim/eval.c [hunk 1]
```diff
@@ -700,6 +703,19 @@ void eval_diff(const char *const origfile, const char *const newfile, const char
   set_vim_var_string(VV_FNAME_OUT, NULL, -1);
 }
 
+// Return list of chunks calculated by user.
+list_T* eval_idifffn(const int idx_orig, const int idx_new, char_u *rem_p_dex)
```
Comment (lewis6991) [inline src/nvim/eval.c:707]: I think `eval_diff_fn` would be a more suitable name. Maybe also rename `eval_diff` to `eval_diff_expr`?

File: src/nvim/diff.c [hunk 2]
```diff
@@ -1104,21 +1185,32 @@ static int diff_file_internal(diffio_T *diffio)
...
+    } else {
+      // Otherwise use 'diffexpr' to generate the diff file.
+      eval_diff(tmp_orig, tmp_new, tmp_diff);
+      return OK;
+    }
   }
   // Use xdiff for generating the diff.
   if (dio->dio_internal) {
-    return diff_file_internal(dio);
+    return diff_file_internal(dio, chunks);
```
Comment (lewis6991) [inline src/nvim/diff.c:1213]: This looks like it would break if `diffopt` doesn't have `internal` and `diffexpr=->...`?

You would solve this by returning straight after running `eval_idifffn`.

**Review by lewis6991 — CHANGES_REQUESTED** [review]

**Comment by iago-lito** [comment]
> The title is a bit wrong. This doesn't impact `vim.diff`, it simply allows `diffexpr` to work completely in process and allows it to use `vim.diff` instead of being expected to run an external process.

Agreed. It was changed by @clason to match the commits naming standards I guess. I'm fixing this with explicit ref to #17210.

**Comment by clason** [comment]
Yes, that was me, because the original PR title was singularly undescriptive. (The fact that I guessed wrong should be an indication why the title was bad.)

File: runtime/doc/diff.txt [hunk 3]
```diff
@@ -354,13 +354,21 @@ The "1a2" item appends the line "bbb".
...
-When 'diffexpr' is not empty, Vim evaluates it to obtain a diff file in the
-format mentioned.  These variables are set to the file names used:
+When 'diffexpr' is a non-empty expresion, Vim evaluates it to obtain a diff file
+in the format mentioned.  These variables are set to the file names used:
 
 	v:fname_in		original file
 	v:fname_new		new version of the same file
 	v:fname_out		where to write the resulting diff file
 
+When 'diffexpr' is an expression prefixed with '->', Vim evaluates it to obtain
```
[the remark this answers is not in the document]
Comment (iago-lito) [reply runtime/doc/diff.txt:364]: Cool. I did this because I wasn't sure how to pass `bufnr_in` and `bufnr_new` as arguments to the ref-ed function, and how to retrieve `diff_chunks` value in return. I'll investigate into `call_func_retlist` :)

File: src/nvim/diff.c [hunk 4]
```diff
@@ -878,14 +878,32 @@ static void diff_try_update(diffio_T *dio, int idx_orig, exarg_T *eap)
...
 }
 
+///
+/// Return true if the diffexpr option is set with '->' prefix,
+/// so it means that internal diff is used and customized.
+///
+int diffexpr_internal_prefix(void)
+{
+  char_u *c = p_dex;
+  if (*c == '-') {
```
[the remark this answers is not in the document]
Comment (iago-lito) [reply src/nvim/diff.c:888]: Yupe. FWIW it's my first time writing C code, I'm not familiar with idioms ;)

File: src/nvim/diff.c [hunk 5]
```diff
@@ -1074,6 +1092,69 @@ static int diff_file_internal(diffio_T *diffio)
   memset(&emit_cfg, 0, sizeof(emit_cfg));
   memset(&emit_cb, 0, sizeof(emit_cb));
 
+  if (diffexpr_internal_prefix()) {
```
[the remark this answers is not in the document]
Comment (iago-lito) [reply src/nvim/diff.c:1095]: Okay, I'll refactor this. This is (yet) another consequence of me not being comfortable with `call_func_retlist` and having fallen back to `eval_to_bool` instead.

**Comment by iago-lito** [comment]
@lewis6991 Thank you for the review. I'm fixing this as soon as I get free time again, marking your various points as "resolved" as I proceed :)

**Comment by lewis6991** [comment]
You can squash when it is ready. For now just add commits.

File: src/nvim/diff.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (iago-lito) [reply src/nvim/diff.c:1095]: Hm, I cannot do *everything* within `eval_diff_fn` at least without messing with include headers, because it seems that `diffio_T` or `xdiff_out` are not available within `eval.c`. I'll move this "translation" process to a dedicated function, then call it at least from `diff_file` for now.

File: src/nvim/diff.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (lewis6991) [reply src/nvim/diff.c:888]: Me neither, I just knew a specific example for this when working on the highlight logic recently.

File: src/nvim/diff.c (hunk 2 above)
Comment (iago-lito) [reply src/nvim/diff.c:1213]: Yupe, that's confusion on my side. I thought that `diff_file` was supposed to both fill up `dio` data then display the chunks, but it's only the first and the second is rather `diff_read`. So yes, I can return earlier.

**Comment by iago-lito** [comment]
@lewis6991 Here is a fix integrating most of your feedback. The only thing I did not address was to "do everything within `eval_diff`"`_fn`, because I [supposed](https://github.com/neovim/neovim/pull/17398#discussion_r810324802) it would trigger unecessary headers dance?

File: src/nvim/eval.c [hunk 6]
```diff
@@ -700,6 +702,18 @@ void eval_diff(const char *const origfile, const char *const newfile, const char
...
+// Return list of chunks calculated by user.
+list_T *eval_diff_fn(const int idx_orig, const int idx_new, char_u *fn_name)
+{
+  typval_T args[2];
+  args[0].v_type = VAR_NUMBER;
+  args[1].v_type = VAR_NUMBER;
+  args[0].vval.v_number = idx_orig;
+  args[1].vval.v_number = idx_new;
+  list_T *chunks = call_func_retlist(fn_name, 2, args);
+  return chunks;
```
Comment (lewis6991) [inline src/nvim/eval.c:714]:
```suggestion
  return call_func_retlist(fn_name, 2, args);
```

File: src/nvim/diff.c [hunk 7]
```diff
@@ -1101,20 +1111,104 @@ static int diff_file_internal(diffio_T *diffio)
...
 {
   char *tmp_orig = (char *)dio->dio_orig.din_fname;
   char *tmp_new = (char *)dio->dio_new.din_fname;
   char *tmp_diff = (char *)dio->dio_diff.dout_fname;
   if (*p_dex != NUL) {
-    // Use 'diffexpr' to generate the diff file.
-    eval_diff(tmp_orig, tmp_new, tmp_diff);
-    return OK;
+    // Evaluate function with `in_bufnr` and `new_bufnr` arguments.
+    if (diffexpr_internal_prefix()) {
```
Comment (lewis6991) [inline src/nvim/diff.c:1198]:
```suggestion
    if (STRNCMP(p_dex, "->", 2) == 0) {
```

File: src/nvim/diff.c [hunk 8]
```diff
@@ -1101,20 +1111,104 @@ static int diff_file_internal(diffio_T *diffio)
...
+  success = OK;
+theend:
+  tv_list_unref(chunks);
+  return success;
+}
+
 /// Make a diff between files "tmp_orig" and "tmp_new", results in "tmp_diff".
 ///
 /// @param dio
+/// @param idx_orig Buffer number of original file.
```
Comment (lewis6991) [inline src/nvim/diff.c:1187]: Mention these arguments are only for use with diffexpr

File: src/nvim/diff.c [hunk 9]
```diff
@@ -1101,20 +1111,104 @@ static int diff_file_internal(diffio_T *diffio)
...
   if (*p_dex != NUL) {
-    // Use 'diffexpr' to generate the diff file.
-    eval_diff(tmp_orig, tmp_new, tmp_diff);
-    return OK;
+    // Evaluate function with `in_bufnr` and `new_bufnr` arguments.
+    if (diffexpr_internal_prefix()) {
+      // Ignore the first two '->' characters of the option,
+      // they are not part of the function identifier.
+      char_u *fn_name = p_dex + 2;
+      list_T* chunks = eval_diff_fn(idx_orig, idx_new, fn_name);
```
Comment (lewis6991) [inline src/nvim/diff.c:1202]:
```suggestion
      list_T* chunks = eval_diff_fn(idx_orig, idx_new, p_dex + 2);
```

File: src/nvim/diff.c [hunk 10]
```diff
@@ -1101,20 +1111,104 @@ static int diff_file_internal(diffio_T *diffio)
...
+  listitem_T *n;
+  typval_T *c_val;
+  typval_T *n_val;
+  int i_size;
+  long start_a;
+  long count_a;
+  long start_b;
+  long count_b;
+  int success = FAIL;
+  if (c != NULL) {
```
Comment (lewis6991) [inline src/nvim/diff.c:1133]:
```suggestion
  if (c == NULL) {
    return FAIL;
  }
```

File: src/nvim/diff.c [hunk 11]
```diff
@@ -1101,20 +1111,104 @@ static int diff_file_internal(diffio_T *diffio)
...
-    // Use 'diffexpr' to generate the diff file.
-    eval_diff(tmp_orig, tmp_new, tmp_diff);
-    return OK;
+    // Evaluate function with `in_bufnr` and `new_bufnr` arguments.
+    if (diffexpr_internal_prefix()) {
+      // Ignore the first two '->' characters of the option,
+      // they are not part of the function identifier.
+      char_u *fn_name = p_dex + 2;
+      list_T* chunks = eval_diff_fn(idx_orig, idx_new, fn_name);
+      if (chunks == NULL) {
```
Comment (lewis6991) [inline src/nvim/diff.c:1203]: This check is already done inside `translate_chunks`

File: src/nvim/diff.c [hunk 12]
```diff
@@ -1101,20 +1111,104 @@ static int diff_file_internal(diffio_T *diffio)
...
+  int i_size;
+  long start_a;
+  long count_a;
+  long start_b;
+  long count_b;
+  int success = FAIL;
+  if (c != NULL) {
+    for (; c != NULL; c = TV_LIST_ITEM_NEXT(chunks, c)) {
+      c_val = TV_LIST_ITEM_TV(c);
+      if (c_val->v_type == VAR_LIST) {
```
Comment (lewis6991) [inline src/nvim/diff.c:1136]:
```suggestion
      if (c_val->v_type != VAR_LIST) {
        goto thend;
      }
```

Don't add an else.

File: src/nvim/diff.c [hunk 13]
```diff
@@ -1101,20 +1111,104 @@ static int diff_file_internal(diffio_T *diffio)
...
+        i = c_val->vval.v_list;
+        i_size = tv_list_len(i);
+        if (i_size == 4) {
```
Comment (lewis6991) [inline src/nvim/diff.c:1139]: Same here, you get the point. Do early returns/breaks to minimize nesting.

**Review by lewis6991 — CHANGES_REQUESTED** [review]
Looking a lot better.

You need to also add a testcase in `test/functional`. Take a look at `test/functional/lua/xdiff_spec.lua` and `test/functional/ui/diff_spec.lua` as examples.

To run the testcase:

```
make functionaltest TEST_FILE=test/functional/<YOUR FILE HERE>
```

File: src/nvim/diff.c (hunk 7 above)
Comment (iago-lito) [reply src/nvim/diff.c:1198]: Hm, this would duplicate the hardcoded definition of the `->` prefix. For now it is used once on line [1198](https://github.com/iago-lito/neovim/blob/70d78aef1516022c5d4b101508fa59e6bdb16b40/src/nvim/diff.c#L1198) and once on line [898](https://github.com/iago-lito/neovim/blob/70d78aef1516022c5d4b101508fa59e6bdb16b40/src/nvim/diff.c#L898). I'd find it more DRY that the check be factorized this way in a dedicated function like `diffexpr_internal_prefix`.. maybe you'd prefer a different name?

File: src/nvim/diff.c (hunk 9 above)
Comment (iago-lito) [reply src/nvim/diff.c:1202]: [Similarly](https://github.com/neovim/neovim/pull/17398#discussion_r810481378), I did not like seeing `pdex + 2` appearing twice (once in [argument](https://github.com/iago-lito/neovim/blob/70d78aef1516022c5d4b101508fa59e6bdb16b40/src/nvim/diff.c#L1202) to `eval_diff_fn`, once in [error message](https://github.com/iago-lito/neovim/blob/70d78aef1516022c5d4b101508fa59e6bdb16b40/src/nvim/diff.c#L1204)).
.. but maybe I'm missing one tradeoff here. Is it important, for some other reason, to keep the number of variables/functions low in nvim sources?

Comment (lewis6991) [reply src/nvim/diff.c:1202]: Once you get more comfortable with the codebase, you'll find this more of an idiom. The comment does more than enough to explain what is happening here.

File: src/nvim/diff.c (hunk 7 above)
Comment (lewis6991) [reply src/nvim/diff.c:1198]: Ok fair enough. Maybe `diffexpr_is_fn()` would be better?

File: src/nvim/diff.c (hunk 10 above)
Comment (iago-lito) [reply src/nvim/diff.c:1133]: Well, [IIUC](https://github.com/iago-lito/neovim/blob/70d78aef1516022c5d4b101508fa59e6bdb16b40/src/nvim/eval.c#L1184) this would leak `chunks`, right? But I get the idea: I'll just `goto theend;` instead of `return FAIL;`.

Comment (lewis6991) [reply src/nvim/diff.c:1133]: Ah yes, didn't see the unref.

File: src/nvim/diff.c (hunk 11 above)
Comment (iago-lito) [reply src/nvim/diff.c:1203]: Very true. Well, I'll rather remove it from `translate_chunks` so I can keep the `pdex + 2` logic located here.

**Comment by iago-lito** [comment]
Here are the suggested fixes. I'll get familiar with the testing proces then offer tests in a next commit :)

**Comment by iago-lito** [comment]
Hello @lewis6991, I have been struggling with tests because of a weird behaviour that I do not get. Maybe you or someone here can help me understand what's going on here?
My problem boils down to these few edits I offer in [de52eee20](https://github.com/neovim/neovim/pull/17398/commits/de52eee2047550a43158d6ec3304bd723aaf9104) (temporary commit). In this commit, I set up the new `diffexpr=->MyDiff` feature within a temporary config file prior to testing, and make `MyDiff` do two things:
- Count how many times it has been called.
- Return (only) one dummy chunk, whose length depends on this count.

With this set up, I run the following to watch the tests fail:

    make functionaltest TEST_FILE=test/functional/ui/diff_spec.lua | less

And here is the first failed test I get:
```
test/functional/ui/diff_spec.lua:88: Row 1 did not match.
Expected:
  |*{1:  }{2:------------------}{3:│}{1:  }{4:0                }|
  |*{1:  }^1                 {3:│}{1:  }1                |
  |*{1:  }2                 {3:│}{1:  }2                |
  |*{1:  }3                 {3:│}{1:  }3                |
  |*{1:  }4                 {3:│}{1:  }4                |
  |*{1:  }5                 {3:│}{1:  }5                |
  |*{1:  }6                 {3:│}{1:  }6                |
  |*{1:+ }{5:+--  4 lines: 7···}{3:│}{1:+ }{5:+--  4 lines: 7··}|
  |*{6:~                   }{3:│}{6:~                  }|
  |*{6:~                   }{3:│}{6:~                  }|
  |*{6:~                   }{3:│}{6:~                  }|
  |*{6:~                   }{3:│}{6:~                  }|
  |*{6:~                   }{3:│}{6:~                  }|
  |*{6:~                   }{3:│}{6:~                  }|
  |*{7:<onal-diff-screen-1  }{3:<l-diff-screen-1.2 }|
  |*:set diffopt=filler                     |
Actual:
  |*{1:  }1                 {3:│}{1:  }0                |
  |*{1:  }2                 {3:│}{1:  }1                |
  |*{1:  }{2:------------------}{3:│}{1:  }{4:2                }|
  |*{1:  }{2:------------------}{3:│}{1:  }{4:3                }|
  |*{1:  }3                 {3:│}{1:  }4                |
  |*{7:                                        }|
  |*{UNEXPECTED background = Screen.colors.Red, foreground = Screen.colors.Gray100:opened}                                  |
  |*{UNEXPECTED background = Screen.colors.Red, foreground = Screen.colors.Gray100:diffthis1}                               |
  |*CALLBACK (6)                            |
  |*{UNEXPECTED background = Screen.colors.Red, foreground = Screen.colors.Gray100:diffthis2}                               |
  |*CALLBACK (5)                            |
  |*CALLBACK (4)                            |
  |*CALLBACK (3)                            |
  |*CALLBACK (2)                            |
  |*{UNEXPECTED background = Screen.colors.Red, foreground = Screen.colors.Gray100:before diffopt}                          |
  |*CALLBACK (1)^                            |

To print the expect() call that would assert the current screen state, use
screen:snapshot_util(). In case of non-deterministic failures, use
screen:redraw_debug() to show all intermediate screen states.
```

There are a few things I find puzzling here, and that I do not understand:
1. Why is the callback invoked 6 times?
2. Why does the message `after diffopt` not appear? In general, I have observed that no message appears after the 6th callback, even with direct `semsg` calls from `src/nvim/diff.c`.
3. On the "Actual screen", the chunk spans over 2 lines. But since the callback count is 1, I would expect the chunk size to be 1. In general, I observe that chunks emitted during the last callback are not taken into account, although I think they should be the most relevant.

So, in a nutshell: I cannot find my way to writing consistent tests yet. I suppose that I am missing some relevant part of the logic at play here.. but what is it?
